operator/pkg/controller/cassandradatacenter/cassandradatacenter_controller.go

```diff
@@ -112,6 +119,72 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
+	// Setup watches for Nodes to check for taints being added
+
+	mapFn := handler.ToRequestsFunc(
+		func(a handler.MapObject) []reconcile.Request {
+			log.Info("Node Watch called")
+			requests := []reconcile.Request{}
+
+			nodeName := a.Object.(*corev1.Node).Name
+
+			c := mgr.GetClient()
```

Comment:
So we've generally been quite cautious to avoid hitting the k8s apis in these handlers, in fact, we've gone so far as to maintain back references in metadata to lookup the relevant resources to enqueue requests for so that we don't have to hit the k8s APIs (the elastic operator does something similar). This function will execute every time the most mundane of changes happens to a node. Do we know of any other operator out there that uses the client from one of these EventHandlers?

Reply:
Why are we concerned about using the manager client from inside of a watch handler function?  

The watches are running in the same process ( modulo threading/go-routines ) as the controller itself.  Calling the k8s api from a watcher should be virtually identical as calling the k8s api from the reconcile loop, right?

I am not opposed to maintaining a list of relevant pods in the node metadata, but I am not sure what the concern is.

Reply:
So this might be okay, it's been a while since I've crawled through the controller runtime code. When I've looked into this previously I found that the elasitic operator went out of its way to avoid using the client here, and I've not seen an example of another operator doing this. So if you feel you've done your homework and believe this okay, I'm on board.

One problem you do seem to have is that if you encounter an error for any reason when querying resources, you don't really have a way to try again. So if you fail to list all the pods (due to some transient error, like maybe etcd is under load) then you will effectively lose the event (and you might not get a subsequent one). In reconcile you have the advantage of being able to try again if something goes wrong, but you cannot do that here, at least not as far as I know.

Reply:
I am not sure exactly how to deal with an error in the watcher mapFunction, except to log it.  The worst case scenario is that the cass-operator doesn't notice that a node has the new taint, and the node is subsequently taken down suddenly, taking the pod and possibly a pv with it.  In that case, we are relying on the normal redundancy features of cassandra to handle the loss of the node, which we might need anyway if we can't even get a list of pods from etcd.

If you have a better strategy idea, I'd be happy to give it a go.

Reply:
So errors in the mapping function lead to lost events. Errors during reconciliation lead to a retry. If you build the mapping of nodes to pods (or nodes to datacenters) during reconciliation, you will not have to worry about lost events. Then, in the event handler, you can look at the mapping to find the relevant datacenters without having to do any IO (which might return an error, especially if it is over the network). The dynamic watches does this, but you would probably want a simpler in memory map like how elastic does it.